File: packages/backend/scripts/logs/log_traces.py
import json
from os import environ

import requests
from dotenv import load_dotenv

load_dotenv()

directory_id=environ['directory_id']
app_registration_id=environ['app_registration_id']
secret_value=environ['secret_value']
workspace_id=environ['workspace_id']

response = requests.post(
    f"https://login.microsoftonline.com/{directory_id}/oauth2/v2.0/token",
    data={
        "client_id": app_registration_id,
        "client_secret": secret_value,
        "grant_type": "client_credentials",
        "scope": "https://api.loganalytics.io/.default"
    }
)

token = response.json()["access_token"]

def run_query(label, query):
    result = requests.post(
        f"https://api.loganalytics.io/v1/workspaces/{workspace_id}/query",
        headers={"Authorization": f"Bearer {token}"},
        json={"query": query}
    )
    print(f"\n===** {label} **===")
    print(result.json())

# --- AppTraces ---
run_query(
    "Recent Traces",
    "AppTraces | where TimeGenerated > ago(7d) | order by TimeGenerated desc | limit 20"
)

run_query(
    "Traces by Severity",
    "AppTraces | where TimeGenerated > ago(7d) | summarize count() by SeverityLevel | order by count_ desc"
)

run_query(
    "Traces by Function",
    "AppTraces | where TimeGenerated > ago(7d) | summarize count() by OperationName | order by count_ desc"
)

# run_query(
#     "Exposed Device Keys",
#     "AppTraces | where TimeGenerated > ago(7d) | where Message contains 'deviceKey' | project TimeGenerated, Message, OperationName | order by TimeGenerated desc"
# )

File: packages/backend/scripts/logs/log_traces.ts
import "dotenv/config";
import { writeFileSync, appendFileSync } from "fs";

const outputFile = "log_traces.txt";
writeFileSync(outputFile, "");

function log(line: string): void {
    appendFileSync(outputFile, line + "\n");
}

const directoryId = process.env["directory_id"];
const appRegistrationId = process.env["app_registration_id"];
const secretValue = process.env["secret_value"];

const tokenResponse = await fetch(
    `https://login.microsoftonline.com/${directoryId}/oauth2/v2.0/token`,
    {
        method: "POST",
        headers: { "Content-Type": "application/x-www-form-urlencoded" },
        body: new URLSearchParams({
            client_id: appRegistrationId!,
            client_secret: secretValue!,
            grant_type: "client_credentials",
            scope: "https://api.loganalytics.io/.default"
        })
    }
);

const { access_token: token } = await tokenResponse.json();
const workspaceId = process.env["workspace_id"];

async function runQuery(label: string, query: string): Promise<void> {
    const result = await fetch(
        `https://api.loganalytics.io/v1/workspaces/${workspaceId}/query`,
        {
            method: "POST",
            headers: {
                "Authorization": `Bearer ${token}`,
                "Content-Type": "application/json"
            },
            body: JSON.stringify({ query })
        }
    );

    log(`\n===** ${label} **===`);
    log(JSON.stringify(await result.json(), null, 2));
}

await runQuery(
    "Recent Traces",
    "AppTraces | where TimeGenerated > ago(7d) | order by TimeGenerated desc | limit 20"
);

await runQuery(
    "Traces by Severity",
    "AppTraces | where TimeGenerated > ago(7d) | summarize count() by SeverityLevel | order by count_ desc"
);

await runQuery(
    "Traces by Function",
    "AppTraces | where TimeGenerated > ago(7d) | summarize count() by OperationName | order by count_ desc"
);

await runQuery(
    "Unique Devices from Traces",
    `AppTraces
    | where TimeGenerated > ago(30d)
    | where Message contains 'deviceID'
    | extend deviceID = extract('deviceID.*?([a-f0-9]{64})', 1, Message)
    | where isnotempty(deviceID)
    | summarize dcount(deviceID) by bin(TimeGenerated, 1d)
    | order by TimeGenerated asc`
);
